app/recipe/views.py

```diff
@@ -0,0 +1,104 @@
+from rest_framework import viewsets, mixins, status
+from rest_framework.authentication import TokenAuthentication
+from rest_framework.decorators import action
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+from core.models import Tag, Ingredient, Recipe
+
+from recipe import serializers
+
+
+class BaseRecipeAttributeViewSet(viewsets.GenericViewSet,
+                                 mixins.ListModelMixin,
+                                 mixins.CreateModelMixin):
+    """Manage generic recipe attributes in database"""
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
+        """Return the objects limited to the authenticated user"""
+        assigned_only = bool(
+            int(self.request.query_params.get('assigned_only', 0))
+        )
+        queryset = self.queryset
+        if assigned_only:
+            queryset = queryset.filter(recipe__isnull=False)
+        return queryset.filter(user=self.request.user)\
+                       .distinct()\
+                       .order_by('-name')
+
+    def perform_create(self, serializer):
+        """Create a new object for the authenticated user"""
+        serializer.save(user=self.request.user)
+
+
+class TagViewSet(BaseRecipeAttributeViewSet):
+    """Manage tags in database"""
+    queryset = Tag.objects.all()
+    serializer_class = serializers.TagSerializer
+
+
+class IngredientViewSet(BaseRecipeAttributeViewSet):
+    """Manage ingredients in database"""
+    queryset = Ingredient.objects.all()
+    serializer_class = serializers.IngredientSerializer
+
+
+class RecipeViewSet(viewsets.ModelViewSet):
+    """Manage recipes in database"""
+    serializer_class = serializers.RecipeSerializer
+    queryset = Recipe.objects.all()
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    def _params_to_ints(self, stringIDs):
+        """Converts a list of string IDs in integers"""
+        return [int(str_id) for str_id in stringIDs.split(',')]
+
+    def get_queryset(self):
+        """Return the recipe for the authenticated user"""
+        tags = self.request.query_params.get('tags')
+        ingredients = self.request.query_params.get('ingredients')
+        queryset = self.queryset
+        if tags:
+            tag_ids = self._params_to_ints(tags)
+            queryset = queryset.filter(tags__id__in=tag_ids)
```

Comment:
I think we are not checking that the user has access to those tags, do we? Perhaps we should guard against that.

Reply:
I think it's done in the following filter, first it filters the queryset take the tags/ingredients based on the IDs, but then it returns `queryset.filter(user=self.request.user)` (line 71), so it should be OK, don't you agree?

Reply:
I believe that filters the recipes to be accessible by the user, but doesn't filter that the indicated tags and ingredients are accessible by the user, right?